libnetwork/drivers/bridge/setup_ip_forwarding.go

```diff
@@ -12,78 +12,65 @@ import (
-				if err := configureIPForwarding(false); err != nil {
-					log.G(context.TODO()).Errorf("Disabling IP forwarding failed, %v", err)
-				}
-				return err
+	if ipv4ForwardData[0] == '1' {
+		return nil
+	}
+	if err := os.WriteFile(ipv4ForwardConf, []byte{'1', '\n'}, forwardConfPerm); err != nil {
+		return fmt.Errorf("Enabling IP forwarding failed: %v", err)
+	}
+	defer func() {
+		if retErr != nil {
+			if err := os.WriteFile(ipv4ForwardConf, []byte{'0', '\n'}, forwardConfPerm); err != nil {
+				log.G(context.TODO()).WithError(err).Error("Disabling IPv4 forwarding failed")
 			}
-			iptables.OnReloaded(func() {
-				log.G(context.TODO()).Debug("Setting the default DROP policy on firewall reload")
-				if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-					log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
-				}
-			})
 		}
-	}
+	}()
 
-	// add only iptables rules - forwarding is handled by setupIPv6Forwarding in setup_ipv6
-	if enableIP6Tables {
-		iptable := iptables.GetIptable(iptables.IPv6)
-		if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-			log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
+	// When enabling ip_forward set the default policy on forward chain to drop.
+	if wantFilterForwardDrop {
+		if err := setFilterForwardDrop(iptables.IPv4); err != nil {
+			return err
 		}
-		iptables.OnReloaded(func() {
-			log.G(context.TODO()).Debug("Setting the default DROP policy on firewall reload")
-			if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-				log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
-			}
-		})
 	}
-
 	return nil
 }
 
-func setupIPv6Forwarding(config *networkConfiguration, i *bridgeInterface) error {
+func setupIPv6Forwarding(wantFilterForwardDrop bool) (retErr error) {
+	configUpdated := false
+
 	// Get current IPv6 default forwarding setup
+	// FIXME(robmry) - is it necessary to set this, setting "all" (below) does the job?
 	ipv6ForwardDataDefault, err := os.ReadFile(ipv6ForwardConfDefault)
 	if err != nil {
 		return fmt.Errorf("Cannot read IPv6 default forwarding setup: %v", err)
 	}
 	// Enable IPv6 default forwarding only if it is not already enabled
 	if ipv6ForwardDataDefault[0] != '1' {
-		if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'1', '\n'}, ipv6ForwardConfPerm); err != nil {
+		if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'1', '\n'}, forwardConfPerm); err != nil {
 			log.G(context.TODO()).Warnf("Unable to enable IPv6 default forwarding: %v", err)
 		}
+		configUpdated = true
+		defer func() {
+			if retErr != nil {
+				if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'0', '\n'}, forwardConfPerm); err != nil {
+					log.G(context.TODO()).WithError(err).Error("Disabling IPv6 default forwarding failed")
```

Comment:
The `WriteFile` error above is ignored (only logged), but here we try to write to it again in the `defer`. Wondering if the first write should be returned as well (there's nothing to "undo" if we failed to write, and it's also likely that if we failed to write the first time that it'll fail the second time?)

The `configUpdated` also may be incorrect here, because if we failed to write, then no config was updated 🤔

Reply:
Right, yes - I've aligned IPv4/IPv6 error handling as you suggested above.